InterruptServer.rb

```diff
@@ -0,0 +1,233 @@
+require 'socket'
+require 'json'
+
+class InterruptServer
+
+  CHAT_LENGTH = 45 # length of chat text user can see at one time
+  TICK_LENGTH = 2 # how many seconds to wait before 'moving chat text left'
+  MAX_MSG_LENGTH = 1024 # max length of incoming message read
+  
+  def initialize(host, port, num_buckets=5)
+    @server = UDPSocket.new
+    @server.bind(host, port)
+
+    puts "running on host: #{host}"
+
+    @clients = {}
+    @buckets_available = [*0..num_buckets-1]
+    # buckets: at most one client can be assigned a given bucket, though clients don't have to be assigned a bucket
+    # the intended use on the client side is to use this to assign colors to users
+
+    @chat_array = Array.new(CHAT_LENGTH, [' ', nil]) # each element is [character, bucket]
+    @outbox = [] # array of messages, as hashes with sender info
+  end
+
+  def run
```

Comment:
if there is only 1 public instance method on an object, the convention is to use `#call` . I don't know the exact story behind that, but I bet it's got something to do with being able to invoke it this way http://stackoverflow.com/questions/19189632/why-does-ruby-call-the-call-method-when-i-dont-supply-the-method-name